Magento1 doc (#103)
front-commerce/developers.front-commerce.com: source/docs/magento1/add-custom-endpoint.md

---
id: add-custom-endpoint
title: Add your custom endpoint
---

> You can add your custom endpoint easily and fasted.

Author: I understand, it seems logical :)

Follow this main step :

1. Add and complete `api2.xml` file
2. Implement API method(s)

## Add and complete api2.xml

> `api2.xml` is your main api config file, you can add this file on your own module in `etc` directory.

Author: Magento parse all `api2.xml` files for create REST API. Same logic with other config file `config.xml`, `system.xml`...

> `api2.xml` is your main api config file, you can add this file on your own module in `etc` directory.

Author: cf `$config = Mage::getConfig()->loadModulesConfiguration('api2.xml');` `Mage_Api2_Model_Config` line 82

                        <retrieve>1</retrieve>
                        <update>1</update>
                        <delete>1</delete>
                    </customer>
                    <guest>
                        <create>1</create>
                        <retrieve>1</retrieve>
                        <update>1</update>
                        <delete>1</delete>
                    </guest>
                </privileges>
                <attributes>
                    <!-- ATTRIBUTE TO EXPOSE -->
                    <entity_id>Entity ID </entity_id>
                </attributes>
                <routes>
                    <!-- API ROUTES -->
                    <route_entity>
                        <route>/entity/:id</route>
                        <action_type>entity</action_type>
                    </route_entity>
                    <route_collection>
                        <route>/entities</route>
                        <action_type>collection</action_type>
                    </route_collection>
                </routes>
                <versions></versions>
            </{API_UNIQUE_NAME}>
        </resources>
    </api2>
</config>
```

- Node description :

>   - [Optional]`resource_groups` use for declare new resource group. You can retrieve this resource when you update API resource Role. Every resource need to be affected to a resource group

Author: I update that with picture, more easy to understand thanks to that.

- [DELETE] `protected function _delete()` call for `entity` action type, body is mandatory
- [DELETE] `protected function _multiDelete()` call for `collection` action type, body is mandatory

### Example

```
<?php

/**
 * Class Module_Network_Model_Api2_Post_Rest_Guest_V1
 */
class Module_Network_Model_Api2_Post_Rest_Guest_V1 extends FrontCommerce_Integration_Model_Api2_Abstract
{
    /**
     * Get post list
     *
     * @return array
     */
    protected function _retrieveCollection()
    {
        $collection = $this->_getCollection();
        $this->_applyCollectionModifiers($collection);
        $this->_loadCollection($collection);
        $this->addCacheHeaders($collection->getCacheLifetime());
        $data = $collection->walk('toArray');
        return array_values((array) $data);
    }

    /**
     * Retrieve post collection
     *
     */
    protected function _getCollection()
    {
        $collection = Mage::getResourceModel('module_network/post_collection');
        return $collection;
    }

    /**
     * Retrieve information about specified socialize
     *
     * @throws Mage_Api2_Exception
     * @return array
     */
    protected function _retrieve()
    {

>         $entityId = this->getRequest()->getParam('entity_id');

Author: You right, it's id instanceof entity id for this example

        }
        return $post->getData();
    }
}

```

### Testing

Best way for testing your new endpoint is to use Postman (or anything else like that).

- Guest testing : If your endpoint can be access in guest mode, you can simply send GET / POST / DELETE / UPDATE request
  to your endpoint for see the response
- Customer testing : You need to add credential and token for your request. You can retrieve all this informations in your database

  - Consumer Key = `key` in table `oauth_consumer`
  - Consumer Secret = `secret` in table `oauth_consumer`
  - Access Token = `token` of user in table `oauth_token`
  - Token secret = `secret` of same user in table `oauth_token`
    ![Customer test](./assets/testing.png)

## Good to know

- If you can, extend `FrontCommerce_Integration_Model_Api2_Abstract` in your own API class. This class add useful functions.
- If you apply `_applyCollectionModifiers($collection)` to your own `$collection` you can use dynamic API collection filter (see [Magento documention](https://devdocs.magento.com/guides/m1x/api/rest/get_filters.html))
- For implement any API error, use `$this->_critical(ERR_CODE);`

> - If you have complexe API, you can override `dispatch()` method for add more action and operation.

Author: I remove this line, I think it's too complicated to understand and useless to know that when you start create custom API.

---
id: add-custom-endpoint
title: Add your custom endpoint
---

When customizing your shop, you may need to retrieve data from Magento. This is done by fetching data from an REST endpoint on Magento that will be used in a [GraphQL module](/docs/essentials/extend-the-graphql-schema.html). In this documentation, you will learn how to create a new REST endpoint in your Magento 1.

This can be done by completing the following steps:

1. Add and complete `api2.xml` file
2. Implement the API method(s)

## Add and complete `api2.xml`

> `api2.xml` is your main API config file, you can add this file on your local module in `etc` directory.

Reviewer: I tend to add a comment at the beginning of each code example with the full path. However, when I do so it is more likely to be a guide rather than an API reference. Not sure in which case we are here. It's mid way between a guide and a reference page.

If we are adding full module paths (which is better imo), we should add also context at the beginning of the page explaining that for the examples shown below, we are creating an endpoint that manages an example resource (with a concrete example).

> `api2.xml` is your main API config file, you can add this file on your local module in `etc` directory.

Author: I use module name on the example step, here this is for explain the different node.

                    </customer>
                    <guest>
                        <retrieve>1</retrieve>
                    </guest>
                </privileges>
                <attributes>
                    <entity_id>Post ID</entity_id>
                    <title>Title</title>
                    <descrition>Description</descrition>
                    <link>Link</link>
                    <image>Image</image>
                </attributes>
                <routes>
                    <route_entity>
                        <route>/social-network-post/:id</route>
                        <action_type>entity</action_type>
                    </route_entity>
                    <route_collection>
                        <route>/social-network-posts</route>
                        <action_type>collection</action_type>
                    </route_collection>
                </routes>
                <versions></versions>
            </social_network_posts>
        </resources>
    </api2>
</config>
```

## Implement API methods

### Directory
First, your file structure should look like this:
`[MODULE]/[MODEL]/Rest/[Guest / Customer / Admin]/V1.php`

> For the example with the social network, the directory is :

Reviewer 2: ```suggestion
In the social network example above, the directory would be:
```

- `[MODULE]/Model/Api2/Post/Rest/Guest/V1.php` (for guest mode)
- `[MODULE]/Model/Api2/Post/Rest/Customer/V1.php` (for customer mode)

This file is your API entrypoint.

<blockquote class="warning">

>   **Warning:** Never forget to add Customer endpoint. If it is the same as the Guest endpoint, you can extend `Guest/V1.php` in `Customer/V1.php`. But if you don't do it, the logged in users won't be able to fetch data from your endpoint, breaking your feature once logged in.

Reviewer 2: ```suggestion
  **Warning:** never forget to add Customer endpoint. If it is the same as the Guest endpoint, you can extend `Guest/V1.php` in `Customer/V1.php`. But if you don't do so, logged in users won't be able to fetch data from Magento, breaking your feature once logged in.
```

### Example

```
<?php

/**
 * Class Module_Network_Model_Api2_Post_Rest_Guest_V1
 */
class Module_Network_Model_Api2_Post_Rest_Guest_V1 extends FrontCommerce_Integration_Model_Api2_Abstract
{
    /**
     * Get post list
     *
     * @return array
     */
    protected function _retrieveCollection()
    {
        $collection = $this->_getCollection();
        $this->_applyCollectionModifiers($collection);
        $this->_loadCollection($collection);
        $this->addCacheHeaders($collection->getCacheLifetime());
        $data = $collection->walk('toArray');
        return array_values((array) $data);
    }

    /**
     * Retrieve post collection
     *
     */
    protected function _getCollection()
    {
        $collection = Mage::getResourceModel('module_network/post_collection');
        return $collection;
    }

    /**
     * Retrieve information about specified socialize
     *
     * @throws Mage_Api2_Exception
     * @return array
     */
    protected function _retrieve()
    {
        $entityId = this->getRequest()->getParam('id'); // param name is defined on your route node, for this example is :id

>         /* @var $stockItem Mage_CatalogInventory_Model_Stock_Item */

Reviewer 2: Oops :)

        $post = Mage::getModel('module_network/post')->load($entityId);
        if (!$post || !$post->getId() || $post->getId() != $entityId) {
            $this->_critical(self::RESOURCE_NOT_FOUND);
        }
        return $post->getData();
    }
}

```

> ### Testing

Reviewer 2: ```suggestion
### Sending requests to your API
```

;)

> In case you need to test your endpoints, you can either use cURL or [Postman](https://www.getpostman.com/). But advanced clients like Postman will allow you to have credentials and will make it easier to test customers vs guest endpoints.

Reviewer 2: ```suggestion
In case you need to ensure your endpoints works as expected, you can either use cURL or [Postman](https://www.getpostman.com/). But advanced clients like Postman will allow you to have credentials and will make it easier to run requests as logged in customers.
```

> - Guest testing : If your endpoint can be access in guest mode, you can simply send GET / POST / DELETE / UPDATE request

Reviewer 2: FYI there are no "spaces" before punctuation in English!

```suggestion
- Guest testing: If your endpoint can be access in guest mode, you can simply send GET / POST / DELETE / UPDATE request
```

  to your endpoint for see the response
- Customer testing : You need to add credential and token for your request. You can retrieve all this informations in your database

  - Consumer Key = `key` in table `oauth_consumer`
  - Consumer Secret = `secret` in table `oauth_consumer`
  - Access Token = `token` of user in table `oauth_token`
  - Token secret = `secret` of same user in table `oauth_token`
    ![The Authorization tab of a Postman request with OAuth 1.0 selected.](./assets/testing.png)

## Good to know

> - If you can, extend `FrontCommerce_Integration_Model_Api2_Abstract` in your own API class. This class add useful functions like:

Reviewer 2: ```suggestion
- If you can, extend `FrontCommerce_Integration_Model_Api2_Abstract` in your own API class. This class adds useful helper functions such as:
```

    - `public function getCustomer()` : Retrieve current customer and save it in customer session.
    - `protected function _initStore()` : Set current store with default store view or store set in request params.

>     - `protected function _getStore()` : Rewrite Magento get store base method for no repeat function if store was already set. Retrieve current store according to request and API user type.

Reviewer 2: I am not sure my understanding is correct, but I suggest the following sentence (easier to read IMO):

```suggestion
    - `protected function _getStore()` : Rewrite Magento's base method to memoize store value. Retrieve current store according to request and API user type.
```

    - `protected function _getCurrency()` : Retrieve current currency.
    - ...

> - If you apply `_applyCollectionModifiers($collection)` to your own `$collection` you can use dynamic API collection filter (see [Magento documention](https://devdocs.magento.com/guides/m1x/api/rest/get_filters.html))

Reviewer 2: ```suggestion
- If you apply `_applyCollectionModifiers($collection)` to your own `$collection` you can use dynamic API collection filter (see [Magento documentation](https://devdocs.magento.com/guides/m1x/api/rest/get_filters.html))
```
